### DISCUSSION TITLE
XRDP connection

### DISCUSSION THREAD

**Discussion by christianstrauch** [notro/gud#18] [Ideas]
All,
I'm contemplating attempting to loop outputs to and inputs from an XRDP session on the Raspberry Pi. The idea is to use a Pi Zero W as a USB client on [any] host, exposing HIDs and graphics via the Pi Zero to an XRDP session and thus creating a universal RDP gadget.
Has anyone ever attempted this or something similar? Any ideas on specific parts of the solution where implementations exist?

**Comment by notro** [comment]
You mean the Pi Zero acting as a USB GUD/HID to XRDP bridge?
The GUD Linux gadget implementation can only connect to a kms device so you would have to write a userspace implementation.

It would have been interesting to see how well a userspace implementation would perform: https://github.com/notro/gud/wiki/Gadget-Implementations#userspace-linux

Hmm maybe you can make use of a writeback connector in order to get the framebuffer content from userspace, if so you could use the vc4 driver: https://forums.raspberrypi.com/viewtopic.php?t=334068#p2000304
The [vkms](https://elixir.bootlin.com/linux/latest/source/drivers/gpu/drm/vkms) virtual test driver also supports a writeback connector. I haven't used a writeback connector so can't help you there.
The problem with this solution, if it works, is that you will only get full screen updates, there's no info about partial updates.

**Reply by christianstrauch** [reply]
Yes, that's what I'm trying to do. I'll experiment with vc4 via userspace. 
I'm wondering about your comment regarding partial updates - wouldn't the XRDP implementation abstract from that?

**Comment by notro** [comment]
You have to memcpy the whole framebuffer instead of just the part that changed, like a mousepointer, which is slower.
